File: scripts/execute-arb.ts
/**
 * Execute arbitrage: buy on one DEX, sell on the other.
 * Uses V0 transaction with Address Lookup Table.
 *
 * Usage:
 *   npx ts-node --skip-project scripts/execute-arb.ts
 *
 * Env:
 *   RPC_URL      — Solana RPC (default: mainnet-beta)
 *   AMOUNT_SOL   — SOL amount to arb (default: 0.001)
 *   ROUTE        — "pump-meteora" or "meteora-pump" (default: pump-meteora)
 *   DRY_RUN      — "true" to only simulate (default: false)
 *   LUT_ADDRESS  — reuse existing LUT (skip creation)
 */

import {
  Connection, PublicKey, ComputeBudgetProgram, AccountMeta,
} from "@solana/web3.js";
import {
  TOKEN_PROGRAM_ID, TOKEN_2022_PROGRAM_ID, ASSOCIATED_TOKEN_PROGRAM_ID,
  getAssociatedTokenAddressSync,
} from "@solana/spl-token";
import { SystemProgram } from "@solana/web3.js";
import * as anchor from "@coral-xyz/anchor";
import * as fs from "fs";
import {
  LIA, WSOL, PUMP, PUMP_FEE, PUMP_POOL, PUMP_GLOBAL_CONFIG,
  PUMP_EVENT_AUTHORITY, PUMP_COIN_CREATOR_VAULT_ATA, PUMP_COIN_CREATOR_VAULT_AUTHORITY,
  PUMP_FEE_CONFIG, PUMP_STATIC_ACCOUNT, PUMP_GLOBAL_VOLUME_ACC,
  PUMP_PROTOCOL_FEE_RECIPIENT_BUY, PUMP_PROTOCOL_FEE_RECIPIENT_ATA_BUY,
  PUMP_PROTOCOL_FEE_RECIPIENT_SELL, PUMP_PROTOCOL_FEE_RECIPIENT_ATA_SELL,
  PUMP_POOL_BASE_VAULT, PUMP_POOL_QUOTE_VAULT,
  METEORA, METEORA_LB_PAIR, METEORA_RESERVE_X, METEORA_RESERVE_Y,
  METEORA_ORACLE, METEORA_EVENT_AUTH, MEMO,
  loadKeypair, deriveUserAccounts, buildSetupIxs,
  getActiveBinArrays, createLookupTable, sendOrSimulateWithLUT,
} from "./utils";

const PROGRAM_ID = new PublicKey("An3HM7PCKigYDszLj8iWYK7mWRnnnhECfM2tRZwsBFV9");

async function main() {
  const rpcUrl = process.env.RPC_URL || "https://api.mainnet-beta.solana.com";
  const dryRun = process.env.DRY_RUN === "true";
  const route = (process.env.ROUTE || "pump-meteora").toLowerCase();
  const amountSol = parseFloat(process.env.AMOUNT_SOL || "0.001");
  const amountIn = BigInt(Math.round(amountSol * 1e9));
  const lutAddressEnv = process.env.LUT_ADDRESS;

  const kp = loadKeypair();
  const conn = new Connection(rpcUrl, "confirmed");
  const { userWsol, userLia, userVolumeAcc, userVolumeAccWsol } = deriveUserAccounts(kp.publicKey);

  const [configPda] = PublicKey.findProgramAddressSync([Buffer.from("config")], PROGRAM_ID);
  const [operatorPda] = PublicKey.findProgramAddressSync(
    [Buffer.from("operator"), kp.publicKey.toBuffer()], PROGRAM_ID,
  );

  const buyOnPump = route === "pump-meteora";

  console.log("Wallet:", kp.publicKey.toBase58());
  console.log("Route:", buyOnPump ? "buy PumpFun → sell Meteora" : "buy Meteora → sell PumpFun");
  console.log("Amount:", amountSol, "SOL (", amountIn.toString(), "lamports )");
  console.log("Dry run:", dryRun);

  // ── Fetch bin arrays ───────────────────────────────────────────────────────

  console.log("\nMeteora pool:");
  const binArrays = await getActiveBinArrays(conn);

  // ── Setup instructions ─────────────────────────────────────────────────────

  const setupIxs = await buildSetupIxs(conn, kp.publicKey, true, amountIn);

  // ── Build arb instruction via Anchor ───────────────────────────────────────

  const wallet = new anchor.Wallet(kp);
  const provider = new anchor.AnchorProvider(conn, wallet, { commitment: "confirmed" });
  const idl = JSON.parse(fs.readFileSync("target/idl/solana_arbitrage_program.json", "utf-8"));
  const program = new anchor.Program(idl, provider);

  // PumpFun accounts depend on direction (buy vs sell use different fee recipients)
  const pumpAccounts = buyOnPump
    ? {
        pumpProtocolFeeRecipient: PUMP_PROTOCOL_FEE_RECIPIENT_BUY,
        pumpProtocolFeeRecipientAta: PUMP_PROTOCOL_FEE_RECIPIENT_ATA_BUY,
        pumpExtraAccountA: userVolumeAccWsol,
        pumpExtraAccountB: PUMP_STATIC_ACCOUNT,
      }
    : {
        pumpProtocolFeeRecipient: PUMP_PROTOCOL_FEE_RECIPIENT_SELL,
        pumpProtocolFeeRecipientAta: PUMP_PROTOCOL_FEE_RECIPIENT_ATA_SELL,
        pumpExtraAccountA: userVolumeAccWsol,
        pumpExtraAccountB: userVolumeAcc,
      };

  const routes = buyOnPump
    ? [
        { dex: { pumpFun: {} }, tokenIn: WSOL, tokenOut: LIA },
        { dex: { meteora: {} }, tokenIn: LIA, tokenOut: WSOL },
      ]
    : [
        { dex: { meteora: {} }, tokenIn: WSOL, tokenOut: LIA },
        { dex: { pumpFun: {} }, tokenIn: LIA, tokenOut: WSOL },
      ];

  const arbIx = await program.methods
    .executeArb(routes, new anchor.BN(amountIn.toString()))
    .accountsPartial({
      user: kp.publicKey,
      operator: operatorPda,
      config: configPda,
      baseMint: LIA,
      quoteMint: WSOL,
      userBaseAta: userLia,
      userQuoteAta: userWsol,
      baseTokenProgram: TOKEN_2022_PROGRAM_ID,
      quoteTokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: SystemProgram.programId,
      associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
      // PumpFun
      pumpPool: PUMP_POOL,
      pumpGlobalConfig: PUMP_GLOBAL_CONFIG,
      pumpPoolBaseVault: PUMP_POOL_BASE_VAULT,
      pumpPoolQuoteVault: PUMP_POOL_QUOTE_VAULT,
      pumpEventAuthority: PUMP_EVENT_AUTHORITY,
      pumpProgram: PUMP,
      pumpCoinCreatorVaultAta: PUMP_COIN_CREATOR_VAULT_ATA,
      pumpCoinCreatorVaultAuthority: PUMP_COIN_CREATOR_VAULT_AUTHORITY,
      pumpGlobalVolumeAccumulator: PUMP_GLOBAL_VOLUME_ACC,
      pumpUserVolumeAccumulator: userVolumeAcc,
      pumpFeeConfig: PUMP_FEE_CONFIG,
      pumpFeeProgram: PUMP_FEE,
      pumpStaticAccount: PUMP_STATIC_ACCOUNT,
      ...pumpAccounts,
      // Meteora
      meteoraLbPair: METEORA_LB_PAIR,
      meteoraBinArrayBitmapExtension: METEORA,
      meteoraReserveX: METEORA_RESERVE_X,
      meteoraReserveY: METEORA_RESERVE_Y,
      meteoraOracle: METEORA_ORACLE,
      meteoraHostFeeIn: METEORA, // must be program ID, not userWsol
      meteoraEventAuthority: METEORA_EVENT_AUTH,
      meteoraMemoProgram: MEMO,
      meteoraProgram: METEORA,
    })
    .remainingAccounts(
      binArrays.map((p) => ({ pubkey: p, isWritable: true, isSigner: false } as AccountMeta))
    )
    .instruction();

  // ── Collect all addresses for LUT ──────────────────────────────────────────

  const allAddrs = new Set<string>();
  const lutAddrs: PublicKey[] = [];
  const addAddr = (pk: PublicKey) => {
    const k = pk.toBase58();
    if (!allAddrs.has(k)) { allAddrs.add(k); lutAddrs.push(pk); }
  };

  for (const key of arbIx.keys) addAddr(key.pubkey);
  for (const ix of setupIxs) {
    for (const key of ix.keys) addAddr(key.pubkey);
  }

  // ── Get or create LUT ──────────────────────────────────────────────────────

  let lookupTable;
  if (lutAddressEnv) {
    console.log("\nUsing existing LUT:", lutAddressEnv);
    const result = await conn.getAddressLookupTable(new PublicKey(lutAddressEnv));
    if (!result.value) throw new Error("LUT not found");
    lookupTable = result.value;
  } else {
    console.log("\nCreating LUT with", lutAddrs.length, "addresses...");
    lookupTable = await createLookupTable(conn, kp, lutAddrs);
  }

  // ── Send ───────────────────────────────────────────────────────────────────

  const allIxs = [
    ComputeBudgetProgram.setComputeUnitLimit({ units: 600_000 }),
    ...setupIxs,
    arbIx,
  ];

  await sendOrSimulateWithLUT(conn, allIxs, kp, lookupTable, dryRun);

  console.log("\nLUT address (reuse with LUT_ADDRESS env):", lookupTable.key.toBase58());
}

main().catch((e) => { console.error("Fatal:", e.message); process.exit(1); });

File: programs/solana_arbitrage_program/src/instructions/execute_arb.rs
use anchor_lang::prelude::*;

use crate::cpi::meteora::{self, MeteoraAccounts};
use crate::cpi::pumpfun::{self, PumpFunAccounts};
use crate::errors::ArbError;

use super::accounts::ExecuteArb;

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq)]
pub enum DexType {
    PumpFun = 0,
    Meteora = 1,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
pub struct Route {
    pub dex: DexType,
    pub token_in: Pubkey,
    pub token_out: Pubkey,
}

impl<'info> ExecuteArb<'info> {
    pub fn process(
        ctx: Context<'_, '_, '_, 'info, ExecuteArb<'info>>,
        routes: Vec<Route>,
        amount_in: u64,
    ) -> Result<()> {
        require!(routes.len() >= 2, ArbError::RouteTooShort);
        validate_route_chain(&routes)?;

        let wsol_mint = ctx.accounts.user_quote_ata.mint;
        require!(
            routes[0].token_in == wsol_mint
                && routes[routes.len() - 1].token_out == wsol_mint,
            ArbError::InvalidRouteTokens
        );

        // TODO: re-enable profit check for production
        // let wsol_before = ctx.accounts.user_quote_ata.amount;

        let mut current_amount = if amount_in == 0 {
            ctx.accounts.user_quote_ata.amount
        } else {
            amount_in
        };

        for route in routes.iter() {
            let is_buy = route.token_in == wsol_mint;

            match route.dex {
                DexType::PumpFun => {
                    let accounts = build_pump_accounts(&ctx.accounts);
                    if is_buy {
                        pumpfun::buy(&accounts, current_amount, 1)?;
                    } else {
                        pumpfun::sell(&accounts, current_amount, 0)?;
                    }
                }
                DexType::Meteora => {
                    let accounts = build_meteora_accounts(&ctx.accounts, is_buy);
                    meteora::swap(&accounts, ctx.remaining_accounts, current_amount, 0)?;
                }
            }

            // Reload both ATAs after each CPI to get fresh balances
            ctx.accounts.user_base_ata.reload()?;
            ctx.accounts.user_quote_ata.reload()?;

            current_amount = if is_buy {
                ctx.accounts.user_base_ata.amount
            } else {
                ctx.accounts.user_quote_ata.amount
            };
        }

        // TODO: re-enable profit check for production
        // ctx.accounts.user_quote_ata.reload()?;
        // require!(ctx.accounts.user_quote_ata.amount > wsol_before, ArbError::NoProfit);

        Ok(())
    }
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

fn validate_route_chain(routes: &[Route]) -> Result<()> {
    for i in 0..routes.len() - 1 {
        require!(routes[i].token_out == routes[i + 1].token_in, ArbError::InvalidRoute);
    }
    Ok(())
}

fn build_pump_accounts<'info>(ctx: &ExecuteArb<'info>) -> PumpFunAccounts<'info> {
    PumpFunAccounts {
        pool: ctx.pump_pool.to_account_info(),
        user: ctx.user.to_account_info(),
        global_config: ctx.pump_global_config.to_account_info(),
        base_mint: ctx.base_mint.to_account_info(),
        quote_mint: ctx.quote_mint.to_account_info(),
        user_base_ata: ctx.user_base_ata.to_account_info(),
        user_quote_ata: ctx.user_quote_ata.to_account_info(),
        pool_base_vault: ctx.pump_pool_base_vault.to_account_info(),
        pool_quote_vault: ctx.pump_pool_quote_vault.to_account_info(),
        protocol_fee_recipient: ctx.pump_protocol_fee_recipient.to_account_info(),
        protocol_fee_recipient_ata: ctx.pump_protocol_fee_recipient_ata.to_account_info(),
        base_token_program: ctx.base_token_program.to_account_info(),
        quote_token_program: ctx.quote_token_program.to_account_info(),
        system_program: ctx.system_program.to_account_info(),
        associated_token_program: ctx.associated_token_program.to_account_info(),
        event_authority: ctx.pump_event_authority.to_account_info(),
        pump_program: ctx.pump_program.to_account_info(),
        coin_creator_vault_ata: ctx.pump_coin_creator_vault_ata.to_account_info(),
        coin_creator_vault_authority: ctx.pump_coin_creator_vault_authority.to_account_info(),
        global_volume_accumulator: ctx.pump_global_volume_accumulator.to_account_info(),
        user_volume_accumulator: ctx.pump_user_volume_accumulator.to_account_info(),
        fee_config: ctx.pump_fee_config.to_account_info(),
        fee_program: ctx.pump_fee_program.to_account_info(),
        extra_account_a: ctx.pump_extra_account_a.to_account_info(),
        extra_account_b: ctx.pump_extra_account_b.to_account_info(),
        static_account: ctx.pump_static_account.to_account_info(),
    }
}

fn build_meteora_accounts<'info>(ctx: &ExecuteArb<'info>, is_buy: bool) -> MeteoraAccounts<'info> {
    let (user_token_in, user_token_out) = if is_buy {
        (ctx.user_quote_ata.to_account_info(), ctx.user_base_ata.to_account_info())
    } else {
        (ctx.user_base_ata.to_account_info(), ctx.user_quote_ata.to_account_info())
    };
    MeteoraAccounts {
        lb_pair: ctx.meteora_lb_pair.to_account_info(),
        bin_array_bitmap_extension: ctx.meteora_bin_array_bitmap_extension.to_account_info(),
        reserve_x: ctx.meteora_reserve_x.to_account_info(),
        reserve_y: ctx.meteora_reserve_y.to_account_info(),
        user_token_in,
        user_token_out,
        token_x_mint: ctx.base_mint.to_account_info(),
        token_y_mint: ctx.quote_mint.to_account_info(),
        oracle: ctx.meteora_oracle.to_account_info(),
        host_fee_in: ctx.meteora_host_fee_in.to_account_info(),
        user: ctx.user.to_account_info(),
        token_x_program: ctx.base_token_program.to_account_info(),
        token_y_program: ctx.quote_token_program.to_account_info(),
        memo_program: ctx.meteora_memo_program.to_account_info(),
        event_authority: ctx.meteora_event_authority.to_account_info(),
        meteora_program: ctx.meteora_program.to_account_info(),
    }
}
